src/python/content.cpp

```diff
@@ -1710,6 +1710,82 @@ make_NumpyArray(const py::handle& m, const std::string& name) {
       .def("simplify", [](const ak::NumpyArray& self) {
         return box(self.shallow_simplify());
       })
+
+      .def_static("from_cupy", [name](py::object array,
+                                                 const py::object& identities,
+                                                 const py::object& parameters) -> ak::NumpyArray {
+        if(py::isinstance(array, py::module::import("cupy").attr("ndarray"))) {
+          std::vector<int64_t> shape, strides;
+
+          shape = pytuples_to_vector<int64_t>(array.attr("shape"));
+          strides = pytuples_to_vector<int64_t>(array.attr("strides"));
+
+          if (py::cast<int64_t>(array.attr("ndim")) == 0) {
+            throw std::invalid_argument(
+              "NumpyArray must not be scalar; try array.reshape(1)");
+          }
+          if (shape.size() != py::cast<int64_t>(array.attr("ndim"))  ||
+              strides.size() != py::cast<int64_t>(array.attr("ndim"))) {
+            throw std::invalid_argument(
+              "CuPy Array len(shape) != ndim or len(strides) != ndim");
+          }
+
+          void* ptr = reinterpret_cast<void *>(py::cast<ssize_t>
+              (array.attr("data").attr("ptr")));
+
+
+          ak::util::dtype cupy_dtype= ak::util::name_to_dtype(py::cast<std::string>
+                                                              (py::str(py::dtype(array.attr("dtype")))));
+
+          return ak::NumpyArray(
+            unbox_identities_none(identities),
+            dict2parameters(parameters),
+            std::shared_ptr<void>(ptr,
+                                  pyobject_deleter<void>(array.ptr())),
+            shape,
+            strides,
+            0,
+            py::dtype(array.attr("dtype")).itemsize(),
+            ak::util::dtype_to_format(cupy_dtype),
+            cupy_dtype,
+            ak::kernel::lib::cuda);
+        }
+        else {
+          throw std::invalid_argument(name + std::string(
+            ".from_cupy() can only accept CuPy Arrays!"));
+        }
+    }, py::arg("array"),
```

Comment:
Some platforms are having trouble compiling this. Don't make the function signature different from the other functions that return layout nodes. The others have a formal return type of `py::object` (the Python object corresponding to a NumpyArray, for instance) and use the `box` helper function to convert a `std::shared_ptr<Content>` into the Python object.

On platforms where this even compiles, it could be a source of segfaults because this NumpyArray is being allocated on the stack and pybind11 doesn't know that it's supposed to take a `shallow_copy` of it to get a copy on the heap. Following the pattern set by all the other functions that return layout nodes, you should

```c++
return box(std::make_shared<ak::NumpyArray>(... arguments ...));
```

to allocate the NumpyArray directly on the heap (no `shallow_copy` needed), get a reference-counted pointer to it, and then `box` it as a Python object. (In principle, pybind11 can do the boxing by itself, but some functions that can return layout nodes can also return scalars, like `None` or numbers. The `box` function does the conversion manually to handle the special cases, and for uniformity, we use it everywhere in this file.)